drivers/serial/uart_esp32.c

```diff
@@ -218,15 +218,16 @@ static int uart_esp32_configure_pins(const struct device *dev, const struct uart
 			break;
 		gpio_pin_set(tx_dev, cfg->tx.pin, 1);
 		gpio_pin_configure(tx_dev, cfg->tx.pin, GPIO_OUTPUT);
-		esp_rom_gpio_matrix_out(cfg->tx.pin, cfg->tx.signal, 0, 0);
+		esp_rom_gpio_matrix_out(cfg->tx.pin,
+				uart_periph_signal[cfg->uart_num].tx_sig, 0, 0);
 
```

Comment:
Is the UART number here actually something that gets written into hardware registers, or is this done this way just as a HAL limitation?

If it's something in hardware, this is OK.

If it's a HAL limitation, then I think you will have to convert this shim driver over to use node labels. This is the same issue we faced in Nordic SoCs for things like the SPI peripheral.

See this comment in spi_nrfx_spim.c, for example:

```c

/*
 * We use NODELABEL here because the nrfx API requires us to call
 * functions which are named according to SoC peripheral instance
 * being operated on. Since DT_INST() makes no guarantees about that,
 * it won't work.
 */
#define SPIM(idx) DT_NODELABEL(spi##idx)
#define SPIM_PROP(idx, prop) DT_PROP(SPIM(idx), prop)
```

Reply:
@mbolivar-nordic, `esp_rom_gpio_matrix_out` needs pin number and its counter part signal for internal routing, here referenced as `.tx_sig`. Each hardware UART (0,1 and 2) has its own `signal` value, that is available in the `hal` array `uart_periph_signal`.  
Before this change, the signal value as set in macro level: `.signal = U##idx##TXD_OUT_IDX,`.
So, UART0 uses `U0TXD_OUT_IDX`, UART1 `U1TXD_OUT_IDX` and UART2 `U2TXD_OUT_IDX`.  
As you know, `DT_INST_FOREACH_STATUS_OKAY` index is not linked to the UART number (0, 1 or 2), which blocks those signal values to be coded as `U##idx##TXD_OUT_IDX` in the macro. So, having the uart number as peripheral configuration allow using `uart_periph_signal` routing table to retrieve all routing signals. So I consider this as hardware config limitation regarding signal routing. Do you think there is another way of having implemented?

Reply:
@sylvioalves I don't see anything in your response that refers to a hardware register. Am I missing something?

Reply:
It does not write the UART number into hardware register at all. However, it uses UART number to find proper signal routing value to configure the hardware register.

Reply:
Then this should not have been merged; can you please send a follow-up to use the node label style I referred to in my above comment?

Reply:
@mbolivar-nordic Martí, sure I can fix this. However, it is not clear to me how that snippet would help in here. In Nordic SPI implementation, SPI device init is called using the approach below:
```c
#ifdef CONFIG_SPI_0_NRF_SPIM
SPI_NRFX_SPIM_DEVICE(0);
#endif
```
where the uart index is "hard-coded" [0,4]. In this approach, you have the SPI hardware instance properly.

However, if you update the driver init to use `DT_INST_FOREACH_STATUS_OKAY`, hardware driver index (`idx` in macro) won't match anymore. Well, you know that already. 

If the driver index is not correct when using this DT macro, how would the macro (you shared) be helpful to fill `idx` value in `U##idx##TXD_OUT_IDX` accordingly?